packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp

```diff
@@ -120,62 +120,131 @@ createTransposeLocal (const Teuchos::RCP<Teuchos::ParameterList> &params)
   typedef Tpetra::Export<LO, GO, Node> export_type;
 
 #ifdef HAVE_TPETRA_MMM_TIMINGS
-    std::string prefix = std::string("Tpetra ")+ label_ + std::string(": ");
-    using Teuchos::TimeMonitor;
-    Teuchos::RCP<Teuchos::TimeMonitor> MM = Teuchos::rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix + std::string("Transpose Local"))));
+  std::string prefix = std::string("Tpetra ")+ label_ + std::string(": ");
+  using Teuchos::TimeMonitor;
+  Teuchos::RCP<Teuchos::TimeMonitor> MM = Teuchos::rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix + std::string("Transpose Local"))));
 #endif
 
+  // Prebuild the importers and exporters the no-communication way,
+  // flipping the importers and exporters around.
+  RCP<const import_type> myImport;
+  RCP<const export_type> myExport;
+  if (! origMatrix_->getGraph ()->getImporter ().is_null ()) {
+    myExport = rcp (new export_type (*origMatrix_->getGraph ()->getImporter ()));
+  }
+  if (! origMatrix_->getGraph ()->getExporter ().is_null ()) {
+    myImport = rcp (new import_type (*origMatrix_->getGraph ()->getExporter ()));
+  }
+
   //
   // This transpose is based upon the approach in EpetraExt.
   //
+
   size_t numLocalCols = origMatrix_->getNodeNumCols();
   size_t numLocalRows = origMatrix_->getNodeNumRows();
   size_t numLocalNnz  = origMatrix_->getNodeNumEntries();
 
-  // Determine how many nonzeros there are per row in the transpose.
-  Array<size_t> CurrentStart(numLocalCols,0);
-  ArrayView<const LO> localIndices;
-  ArrayView<const Scalar> localValues;
   RCP<const crs_matrix_type> crsMatrix =
     rcp_dynamic_cast<const crs_matrix_type> (origMatrix_);
-  if (crsMatrix == Teuchos::null) {
+
+  RCP<crs_matrix_type> transMatrixWithSharedRows;
+  if (crsMatrix != Teuchos::null) {
+    // The matrix is a CrsMatrix
+    using local_matrix_type = typename crs_matrix_type::local_matrix_type;
+    using row_map_type = typename local_matrix_type::row_map_type::non_const_type;
+    using index_type = typename local_matrix_type::index_type::non_const_type;
+    using values_type = typename local_matrix_type::values_type::non_const_type;
+    using execution_space = typename local_matrix_type::execution_space;
+
+    // Determine how many nonzeros there are per row in the transpose.
+    auto lclMatrix = crsMatrix->getLocalMatrix();
+    auto lclGraph  = lclMatrix.graph;
+
+    using range_type = Kokkos::RangePolicy<LO,execution_space>;
+
+    // Determine how many nonzeros there are per row in the transpose.
+    row_map_type t_rows("transpose_rows", numLocalCols+1);
+    Kokkos::parallel_for("compute_number_of_indices_per_column", range_type(0, numLocalRows),
+      KOKKOS_LAMBDA(const LO row) {
+        auto rowView = lclGraph.rowConst(row);
+        auto length  = rowView.length;
+
+        for (decltype(length) colID = 0; colID < length; colID++) {
+          auto col = rowView(colID);
+          Kokkos::atomic_fetch_add(&t_rows[col], 1);
+        }
+    });
+
+    // Compute offsets
```

Comment:
Tpetra has a function for computing offsets from counts: `Tpetra::Details::computeOffsetsFromCounts` I think.  This should be fine, though.